lib/spack/spack/util/url.py

```diff
@@ -92,151 +68,26 @@ def format(parsed_url):
-        paths = paths[last_abs_component:]
-        if len(paths) == 1:
-            result = urllib.parse.urlparse(paths[0], scheme="file", allow_fragments=False)
-
-            # another subtlety: If the last argument to join() is an absolute
-            # file:// URL component with a relative path, the relative path
-            # needs to be resolved.
-            if result.scheme == "file" and result.netloc:
-                result = urllib.parse.ParseResult(
-                    scheme=result.scheme,
-                    netloc="",
-                    path=posixpath.abspath(result.netloc + result.path),
-                    params=result.params,
-                    query=result.query,
-                    fragment=None,
-                )
-
-            return result.geturl()
-
-    return _join(*paths, **kwargs)
-
-
-def _join(base_url, path, *extra, **kwargs):
-    base_url = urllib.parse.urlparse(base_url)
-    resolve_href = kwargs.get("resolve_href", False)
-
-    (scheme, netloc, base_path, params, query, _) = base_url
-    scheme = scheme.lower()
-
-    path_tokens = [
-        part
-        for part in itertools.chain(
-            _split_all(path),
-            itertools.chain.from_iterable(_split_all(extra_path) for extra_path in extra),
-        )
-        if part and part != "/"
-    ]
-
-    base_path_args = ["/fake-root"]
-    if scheme == "s3":
-        if netloc:
-            base_path_args.append(netloc)
-
-    if base_path.startswith("/"):
-        base_path = base_path[1:]
-
-    base_path_args.append(base_path)
-
-    if resolve_href:
-        new_base_path, _ = posixpath.split(posixpath.join(*base_path_args))
-        base_path_args = [new_base_path]
-
-    base_path_args.extend(path_tokens)
-    base_path = posixpath.relpath(posixpath.join(*base_path_args), "/fake-root")
-
-    if scheme == "s3":
-        path_tokens = [part for part in _split_all(base_path) if part and part != "/"]
-
-        if path_tokens:
-            netloc = path_tokens.pop(0)
-            base_path = posixpath.join("", *path_tokens)
-
-    if sys.platform == "win32":
-        base_path = convert_to_posix_path(base_path)
-
-    return format(
-        urllib.parse.ParseResult(
-            scheme=scheme, netloc=netloc, path=base_path, params=params, query=query, fragment=None
-        )
-    )
+def join(base, url, *, resolve_href: bool = False, **kwargs):
+    """Wrapper around ``urllib.parse.urljoin`` that makes s3:// and gs:// URLs behave like
+    http:// URLs. If resolve_href=True, the behavior is the same as a browser: e.g.
+    https://example.com/a/b + c/d = https://example.com/a/c/d. If resolve_href=False, the
+    behavior is like joining paths: https://example.com/a/b + c/d = https://example.com/a/b/c/d."""
+    uses_netloc = urllib.parse.uses_netloc
+    uses_relative = urllib.parse.uses_relative
+    urllib.parse.uses_netloc = [*uses_netloc, "s3", "gs"]
+    urllib.parse.uses_relative = [*uses_relative, "s3", "gs"]
```

Comment:
It may not be forward compatible, but it's backwards compatible for 22 years. I guess that implies it's unlikely to ever be forward incompatible, given that urllib is not really being developed.

Reply:
Alternatively I could temporarily replace the scheme from s3/gs -> http if the right-hand side does not have a scheme, do the join, and revert the scheme.

Reply:
Between what you have and the scheme replacement, I prefer the way it is. I'd prefer an option (c) that didn't use undocumented properties of a library though. I suggest adding a comment that points out this isn't a documented variable, but that it happens to work.

(feel free to collapse with or without doing that)